File: Number of Good Leaf Nodes Pairs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        count = 0

        def explore(node):
            nonlocal count, distance
            if node is None:
                return {}

            if not node.left and not node.right:
                return {0: 1}

            counter = Counter()
            left_counter = explore(node.left)
            right_counter = explore(node.right)
            for k, v in left_counter.items():
                for k2, v2 in right_counter.items():
                    if k + k2 + 2 <= distance:
                        count += v * v2

                if k + 1 < distance:
                    counter[k + 1] += v

            for k, v in right_counter.items():
                if k + 1 < distance:
                    counter[k + 1] += v

            return counter

        explore(root)
        return count

File: Number of Good Leaf Nodes Pairs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countPairs(TreeNode* root, int distance) {
        int count = 0;
        function<unordered_map<int, int>(TreeNode*)> explore;
        explore = [&explore, &count, &distance] (TreeNode* node) -> unordered_map<int, int> {
            if (!node) return {};

            if (!node->left && !node->right) return {{0, 1}};

            unordered_map<int, int> counter;
            auto left_counter = explore(node->left);
            auto right_counter = explore(node->right);
            for (auto [k, v]: left_counter) {
                for (auto [k2, v2]: right_counter) {
                    if (k+k2+2 <= distance) count += v * v2;
                }

                if (k+1 < distance) counter[k+1] += v;
            }

            for (auto [k, v]: right_counter) {
                if (k+1 < distance) counter[k+1] += v;
            }
            return counter;
        };
        explore(root);
        return count;
    }
};
